microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceDiagnosticsParticipant.java

```diff
@@ -0,0 +1,179 @@
+	}
+
+	private static class FaultToleranceAnnotationValidator extends ASTVisitor {
+
+		private IType type;
+		private Set<String> methodSet;
+		private List<Diagnostic> diagnostics;
+		private JavaDiagnosticsContext context;
+
+		private static Logger LOGGER = Logger.getLogger(FaultToleranceAnnotationValidator.class.getName());
+
+		public FaultToleranceAnnotationValidator(IType type, List<Diagnostic> diagnostics, JavaDiagnosticsContext context) {
+			super();
+			this.type = type;
+			this.methodSet = null;
+			this.diagnostics = diagnostics;
+			this.context = context;
+		}
+
+		@Override
+		public boolean visit(MethodDeclaration node) {
+			try {
+				validateMethod(node, type, diagnostics, context);
+			} catch (JavaModelException e) {
+				LOGGER.log(Level.WARNING, "An exception occured when attempting to validate the fallback method");
+			}
+			super.visit(node);
+			return true;
+		}
+
+		private void validateMethod(MethodDeclaration node, IType type, List<Diagnostic> diagnostics,
+				JavaDiagnosticsContext context) throws JavaModelException {
+			List modifiers = node.modifiers();
+			for (Object modifier : modifiers) {
+				if (modifier instanceof NormalAnnotation) {
+					NormalAnnotation annotation = (NormalAnnotation) modifier;
+					if (FALLBACK_ANNOTATION_ABBR.equals(annotation.getTypeName().getFullyQualifiedName())) {
+						Expression fallbackMethodExpr = getFallbackMethodValue(annotation);
+						if (fallbackMethodExpr != null) {
+							String fallbackMethodName = fallbackMethodExpr.toString();
+							fallbackMethodName = fallbackMethodName.substring(1, fallbackMethodName.length() - 1);
+							if (!getMethodSet().contains(fallbackMethodName)) {
+								IOpenable openable = type.getOpenable();
+								Diagnostic d = context.createDiagnostic(context.getUri(),
+										"The method '" + fallbackMethodName + "' does not exist",
+										context.getUtils().toRange(openable,
+												fallbackMethodExpr.getStartPosition(),
+												fallbackMethodExpr.getLength()),
+										DIAGNOSTIC_SOURCE, FALLBACK_METHOD_DOES_NOT_EXIST);
+								d.setSeverity(DiagnosticSeverity.Error);
+								diagnostics.add(d);
+							}
+						}
+					}
+				}
+			}
+		}
+
+		private Set<String> getMethodSet() throws JavaModelException {
+			if (methodSet == null) {
+				methodSet = Stream.of(type.getMethods()).map(m -> {
+					return m.getElementName();
+				}).collect(Collectors.toUnmodifiableSet());
+			}
+			return methodSet;
+		}
+
+	}
+
+
+
+	/**
+	 * Return the value of fallbackMethod as an Expression from the given NormalAnnotation,
+	 * or null if the annotation doesn't have fallbackMethod
+	 *
+	 * @param annotation The annotation to check
+	 * @return the fallbackMethod as an Expression from the given NormalAnnotation,
+	 * or null if the annotation doesn't have fallbackMethod
+	 */
+	private static Expression getFallbackMethodValue(NormalAnnotation annotation) {
```

Comment:
create a generic static method in AnnotationUtiols#getAnnotationMemberValue (like existing method which is working with IAnnotation, but in your case it uses Annotation.

Reply:
I named the method different from the other one, since a Expression is returned instead of a String (I need to use the beginning and ending offsets of the Expression).